File: backend/components/metiers/administration_publique/services/core/controllers/services_controller.test.py
"""
services_controller.test.py
Tests unitaires ultra avancés pour ServicesController (clé en main, CI/CD ready)
"""
import pytest
from .services_controller import ServicesController

def test_services_controller_init():
    ctrl = ServicesController(options={"mode": "test"})
    assert ctrl.options == {"mode": "test"}
    assert ctrl.get_audit_trail() == []

def test_services_controller_config_and_audit():
    ctrl = ServicesController()
    assert ctrl.init({"version": 1}) is True
    assert ctrl.config == {"version": 1}
    assert ctrl.get_audit_trail()[0]["action"] == 'init'

def test_services_controller_handle_valid():
    ctrl = ServicesController()
    ctrl.init({"version": 2})
    result = ctrl.handle("ACTION", {"foo": "bar"})
    assert result == {"success": True, "action": "ACTION", "payload": {"foo": "bar"}, "config": {"version": 2}}
    assert len(ctrl.get_audit_trail()) == 2
    assert ctrl.get_audit_trail()[1]["action"] == 'handle'

def test_services_controller_handle_invalid():
    ctrl = ServicesController()
    with pytest.raises(ValueError):
        ctrl.handle(None, {})
    assert ctrl.get_audit_trail()[0]["action"] == 'error'

File: backend/components/metiers/administration_publique/services/core/controllers/services_controller.test.js
// services_controller.test.js
// Tests unitaires ultra avancés pour ServicesController (clé en main, CI/CD ready)

const ServicesController = require('./services_controller');

describe('ServicesController', () => {
  it('doit s\'initialiser correctement', () => {
    const ctrl = new ServicesController({ mode: 'test' });
    expect(ctrl.options).toEqual({ mode: 'test' });
    expect(ctrl.getAuditTrail()).toEqual([]);
  });

  it('doit initialiser la config et auditer', () => {
    const ctrl = new ServicesController();
    expect(ctrl.init({ version: 1 })).toBe(true);
    expect(ctrl.config).toEqual({ version: 1 });
    expect(ctrl.getAuditTrail()[0].action).toBe('init');
  });

  it('doit gérer une action valide et auditer', () => {
    const ctrl = new ServicesController();
    ctrl.init({ version: 2 });
    const result = ctrl.handle('ACTION', { foo: 'bar' });
    expect(result).toEqual({ success: true, action: 'ACTION', payload: { foo: 'bar' }, config: { version: 2 } });
    expect(ctrl.getAuditTrail().length).toBe(2);
    expect(ctrl.getAuditTrail()[1].action).toBe('handle');
  });

  it('doit lever une erreur sur une action invalide et auditer', () => {
    const ctrl = new ServicesController();
    expect(() => ctrl.handle(null, {})).toThrow('Invalid action');
    expect(ctrl.getAuditTrail()[0].action).toBe('error');
  });
});
